rakelib/plugins_docs_dependencies.rake

```diff
@@ -0,0 +1,169 @@
+      @from = from
+      @priority = from == :default ? 1 : -1
+    end
+
+    def eql?(other)
+      version == other.version && priority == other.priority
+    end
+
+    def <=>(other)
+      if eql?(other)
+        0
+      else
+        [priority, version] <=> [other.priority, other.version]
+      end
+    end
+
+    def to_hash(hash = {})
+      {
+        "version" => version,
+        "from" => from
+      }
+    end
+
+    def to_s
+      "from:#{from}, version: #{version}"
+    end
+  end
+
+  def measure_execution(label)
+    started_at = Time.now
+    response = yield
+    puts "Execution of label: #{label}, #{Time.now - started_at}s"
+    response
+  end
+
+  def all_plugins
+    measure_execution("Fetch all available plugins on `logstash-plugins`") do
+      LogStash::RakeLib.fetch_all_plugins.delete_if { |name| name =~ /^logstash-mixin-/ }
+    end
+  end
+
+
+  # We us a brute force strategy to get the highest version possible for all the community plugins.
+  # We take each plugin and we add it to the current dependencies and we try to resolve the tree, if it work we
+  # record the version installed.
+  def retrieve_definitions
+    builder = Bundler::Dsl.new
+    gemfile = LogStash::Gemfile.new(File.new(LogStash::Environment::GEMFILE_PATH, "r+")).load
+
+    successful_dependencies = {}
+    failures = {}
+
+    builder.eval_gemfile("bundler file", gemfile.generate())
+    definition = builder.to_definition(LogStash::Environment::LOCKFILE, {})
+    extract_versions(definition, successful_dependencies, :default)
+
+    plugins_to_install = all_plugins - successful_dependencies.keys
+    plugins_to_install, partition_size = optimize_for_successful_resolution(plugins_to_install)
+    measure_execution("batch install of plugins") do
+      batch_install(plugins_to_install, successful_dependencies, failures, partition_size)
+    end
+
+    return [successful_dependencies, failures]
+  end
+
+  # If we found the result of a previous execution we will use the successful plugins result
+  # to order the current plugins, we assume that the plugin that was successful will still be successful.
+  # This help us reduce the number of resolve call and make the generation 3 times faster.
+  def optimize_for_successful_resolution(plugins_to_install)
+    if ::File.exist?(EXPORT_FILE)
+      content = JSON.parse(::File.read(EXPORT_FILE))
+
+      possible_success = []
+      possible_failures = []
+      unknown = []
+
+      plugins_to_install.each do |name|
+        if content["successful"][name].nil?
+          if content["failures"][name].nil?
+            unknown << name
+          else
+            possible_failures << name
+          end
+        else
+          possible_success << name
+        end
+      end
+
+      plugins_to_install = possible_success.concat(possible_failures).concat(unknown)
+      [plugins_to_install, plugins_to_install.size / possible_success.size]
+    else
+      [plugins_to_install, 2]
```

Comment:
This is where the magic happens; I am trying to optimize for successful **batch size**, I am brute forcing my way to reduce the number of resolutions required by installing multiple plugins at each iteration.

Let's take the _else_ condition; this is where you hit if you never run the tool before. I will take the plugins list split it into two groups and try to install them at once and recursively reduce the group size until I can install them. In the worse case I need to do a resolution for each plugins, at best (when successful) I do a resolution for a batch of plugins.

If you run the tool before, I will have the previous run cached. I will sort the plugins into two groups, plugins that were previously successful and the other. I am assuming that plugin that was installed successfully will still be installed successfully, so I will try to do a batch of plugins.


_Note I am recursively calling batch install that will reduce the batch for each iteration_